include/realtime_memory/free_list_resource.h

```diff
@@ -114,6 +118,87 @@ class free_list_resource : public cradle::pmr::identity_equal_resource
             previous_in_list->next = next_in_list;
     }
 
+    // Merges contiguous free blocks, and returns the block that's
+    // closest in size to the the desired block size.
+    void* defragment (std::size_t desired_block_size)
+    {
+        if (first_free == nullptr || first_free->next == nullptr)
+            return nullptr;
+
+        // First we sort the free list so that it's ordered by pointer position (lowest first)
+        while (true)
+        {
+            bool anySwapped = false;
+
+            for (mem_block* p0 = first_free->next, *p1 = first_free, *p2 = nullptr; p0 != nullptr && p1 != nullptr;)
```

Comment:
This looks very much like a bubble sort - did you consider other algorithms?

If we're choosing a simple but non-optimal sort here because we believe defragmentation isn't a case that needs to be optimal, we should probably put a comment to that effect, to document to the wider world we've considered it.

(Even with this approach it seems after the first pass you can guarantee the largest element is at the end so each subsequent passes can stop earlier than the last.)

If you were so inclined, there's an opportunity to unit test the sort and consolidate phases of the algorithm separately and then you could test that the implementation of each never drops any blocks. It couldn't be done via the public API though, it would require the implementation to be exposed separately.

Reply:
Yes, I chose this way of doing it because it was the simplest I could think of, and it meets our needs for the moment: the defrag time is imperceptible to a user (well to me) in our current usages of the resource, where this stage only ever becomes necessary in a non-time-critical scenario.

However, you inspired me to hunt for a better algorithm, and I found that merge sort is fairly easy for singly linked lists, and doesn't require auxiliary memory (as it would for an array). So I think I will try to implement that, in a public header so it can be tested.